### ISSUE TITLE
use nuitka to create standalone executable

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#107]
I just did a quick try, it is easy and seems to work.

Maybe somebody using Mac OS X and Windows could also try nuitka with borg there, see: http://nuitka.net/

```
$ nuitka --python-version=3.4 --standalone borg
```

then see the contents of borg.dist/ directory and esp. the borg.exe there (on linux, that is a linux binary, ignore the strange name).

**Comment by oderwat** [comment]
I did:
```bash
> pip3.4 install -U nuitka
> git clone git@github.com:borgbackup/borg.git
> cd borg
> nuitka --standalone borg

nuitka: can't find '__main__' module in 'borg'
```

**Comment by ThomasWaldmann** [comment]
Ooops, I forgot to commit that ... - wait a sec. :)

**Comment by oderwat** [comment]
Cool! On the first glance it works on osx 10.10.4 (python3 installed with homebrew). Will try that executable later some more.

**Comment by oderwat** [comment]
Humm.. I think it is bad that the executable does not load the modules from it's "own" directory. This makes it hard to install cborg (which is what I called it). It would be cool if one could link the ".exe" (cough) into /usr/local/bin .. but then it does not find the other modules so you need to "cd" first (or add a path or something I did not checked for yet)

**Comment by sb56637** [comment]
Is the final standalone executable something that can be distributed, or will it only work on the system it was compiled on? This would be a very attractive option for me.

**Comment by oderwat** [comment]
I think it is distributable on "the same system". Which means I don't think it will work on other OS releases. I am pretty sure my "borg.dist" folder would run on any 10.10.x OSX system **if** they install OSX-Fuse 3.0.x manually. Afaik it packs all python related code and packages into the dist folder.

I guess that will be similar on windows and linux.

**Comment by ThomasWaldmann** [comment]
@oderwat maybe it is rather dependent on "compatible" libraries. So your OS release may not matter as long as the libs work.

You may not have to install osxfuse as long as you don't want to use borg's fuse functionality, try it.

**Comment by oderwat** [comment]
More of a problem is that the binary version created defective backups for me yesterday. They all had defective files (all just 16 bytes I think it was). I am not sure about the fuse stuff. It may be that you can run it without the dependent libraries at load time. I may try that later... there is to much other stuff to do right now :(

**Comment by ThomasWaldmann** [comment]
https://paste.thinkmo.de/J4oPWfvo#borg-ubuntu-14.04-64bit.tar.gz < as the name says.

Please read the included README and test on other Linux distributions / releases, so we'll see the compatibility. Give feedback here.

**Comment by ThomasWaldmann** [comment]
@oderwat see the directory structure and the symlink in that archive file. ^^

**Comment by oderwat** [comment]
Also:
```
> borg.dist/borg.exe init /Volumes/Borg/test3.borg
Initializing repository at "/Volumes/Borg/test3.borg"
Encryption NOT enabled.
Use the "--encryption=repokey|keyfile|passphrase" to enable encryption.
Synchronizing chunks cache...
Rebuilding archive collection. Known: 0 Repo: 0 Unknown: 0
Merging collection into master chunks cache...
Done.

> borg.dist/borg.exe create -v /Volumes/Borg/test3.borg::first .
Segmentation fault: 11
```

Sadly it may not be so easy to get it working on OSX :(

**Comment by ThomasWaldmann** [comment]
@oderwat can you report that segfault issue to nuitka project? Including some instructions how to reproduce?

**Comment by oderwat** [comment]
I used `pip3 install nuitka` which installed 0.5.13.2. Then I just build the executable and get the segfault on any "create". Sorry, but I don't think that I will spent a lot of time with that topic. I have a bunch of other things going on right now and borg works good enough anyway for me right now.

**Comment by ThomasWaldmann** [comment]
Update: we tried the binaries made on ubuntu 14.04 on fedora and they did not work.

http://bugs.nuitka.net/issue224

**Comment by neutrinus** [comment]
Instead of nuitka, pyinstaller worked for many of my projects, maybe it will work for Borg also?

**Comment by ThomasWaldmann** [comment]
@neutrinus do you use it for python 2.x or 3.x stuff?

**Comment by neutrinus** [comment]
@ThomasWaldmann mainly 2.x, but 3.x has been used internally

**Comment by ThomasWaldmann** [comment]
@neutrinus so for making an installer for py 3.x code like borg, one needs to use their git python3 branch as there is not release yet?

**Comment by ThomasWaldmann** [comment]
Tried nuitka again, but ran into an issue again: http://bugs.nuitka.net/issue246

Now trying pyinstaller, they at least have a 3.0.dev pre-release on pypi now.

**Comment by anarcat** [comment]
should this be closed now that #213 works?

**Comment by ThomasWaldmann** [comment]
Yeah, guess so.
